src/common/datasource/src/file_format/parquet.rs

```diff
@@ -51,6 +54,49 @@ impl FileFormat for ParquetFormat {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct DefaultParquetFileReaderFactory {
+    object_store: ObjectStore,
+}
+
+/// Returns a AsyncFileReader factory
+impl DefaultParquetFileReaderFactory {
+    pub fn new(object_store: ObjectStore) -> Self {
+        Self { object_store }
+    }
+}
+
+impl ParquetFileReaderFactory for DefaultParquetFileReaderFactory {
+    // TODO(weny): Supports [`metadata_size_hint`].
+    // The upstream has a implementation supports [`metadata_size_hint`],
+    // however it coupled with Box<dyn ObjectStore>.
+    fn create_reader(
+        &self,
+        _partition_index: usize,
+        file_meta: FileMeta,
+        _metadata_size_hint: Option<usize>,
+        _metrics: &ExecutionPlanMetricsSet,
+    ) -> DatafusionResult<Box<dyn AsyncFileReader + Send>> {
+        let path = file_meta.location().to_string();
+
+        let object_store = self.object_store.clone();
+
+        let reader = std::thread::spawn(move || {
```

Comment:
question: Could we avoid this blocking call? @v0y4g3r @waynexia 

Is it possible to implement the AsyncFileReader trait to work around this?
- The trait implementation calls `object_store.reader()` and then delegates operations to the returned reader

Reply:
Let me implement a prototype with the lazy initialization 🫡